design/accepted/gc_plus/uncomitted-gc.md

```diff
@@ -0,0 +1,201 @@
+The basis for this is a GC Client (i.e. _Spark_ job) consuming objects information from both lakeFS and directly from the underlying object storage,
+and using this information to determine which objects be deleted from the namespace.
+
+The GC process is composed of 3 main parts:
+1. Listing namespace objects
+2. Listing of lakeFS repository committed objects
+3. Listing of lakeFS repository uncommitted objects
+
+Objects that are found in 1 and are not in 2 or 3 can be safely deleted by the Garbage Collector.
+
+### 1. Listing namespace objects
+
+For large repositories, object listing is a very time-consuming operation - therefore we need to find a way to optimize it.
+The suggested method is to split the repository structure into fixed size (upper bounded) slices.
+These slices can then be scanned independently using multiple workers.
+In addition, taking advantage of the common property of the listing operation, which lists objects in a lexicographical order, we can create the slices in a manner which
+enables additional optimizations on the GC process (read further for details).
+To resolve indexing issues with existing repositories that have a flat filesystem, suggested to create these slices under a `data` prefix. 
+
+![Repository Structure](uncommitted-gc-repo-struct.png)
+
+### 2. Listing of lakeFS repository committed objects
+
+Similar to the way GC works today, use repository meta-ranges and ranges to read all committed objects in the repository. 
+
+### 3. Listing of lakeFS repository uncommitted objects
+
+Creating special meta-ranges and ranges using branches' uncommitted data, we can leverage the already existing GC logic to
+quickly read all uncommitted objects in the repository.
+
+### Required changes by lakeFS
+
+The following are necessary changes in lakeFS in order to implement this proposal successfully.
+
+#### Objects Path Conventions
+
+Uncommitted GC must scan the bucket in order to find objects that are not referenced by lakeFS.
+To optimize this process, suggest the following changes:
+
+1. Divide the repository namespace into time-and-size based slices.
+2. Slice name will be a time based, reverse sorted unique identifier.
+3. LakeFS will create a new slice on a timely basis (for example: hourly) or when it has written < MAX_SLICE_SIZE > objects to the slice.
+4. Each slice will be written by a single lakeFS instance in order to track slice size.
+5. The sorted slices will enable partial scans of the bucket when running the optimized GC.
+
+#### StageObject
+
+The StageObject operation will only be allowed on addresses outside the repository's storage namespace. This way, objects added using this operation are never collected by GC.
+
+#### [Get/Link]PhysicalAddress 
+
+1. GetPhysicalAddress to return a validation token along with the address (or embedded as part of the address).
+2. The token will be valid for a specified amount of time and for a single use.
```

Comment:
Why does it matter that the token is valid for a single use?

Reply:
See: https://github.com/treeverse/lakeFS/issues/4438

Reply:
Also on AWS awslabs/amazon-eks-ami#249 is a recurring issue in AMIs, and a linked issue has someone with a 7 minute time skew on EKS.

Clocks should be synced, but I think we will exact too strong a penalty for failure here. We should add:

1. Some safety mechanism.

   For instance:
   * If when we scan we discover prefixes from the distant future (>1 minute?) then error out and mark some metric for alerting.
   *  Create an additional marker object in every partition once an hour, and scan what the other cluster members put in _theirs_.  This gives an estimate of time skew, which can be monitored.
3. Operational guidance (strongly worded documentation).